### PR TITLE
fix(legacy): don't erase firmware and storage in intermediate firmware ...

### PR DESCRIPTION
**Pull request by prusnak**
... if the storage has been already initialized

Possible fix for https://github.com/trezor/trezor-firmware/issues/1597

### CODE REVIEW

**Review by matejcik — APPROVED** [review]
looks good, as best as I can tell

**Review by kusai90 — COMMENTED** [review]
#

**Review by hiviah — APPROVED** [review]
Looks fine, but I didn't get chance to test it on device.

**Comment by prusnak** [comment]
Switching back to draft - we need to polish the details before merging this one in.

File: legacy/intermediate_fw/trezor.c [hunk 1]
```diff
@@ -32,6 +33,12 @@
 #include "timer.h"
 #include "util.h"
 
+// storage magic
+#define FLASH_META_START 0x08008000
+static const uint32_t NORCOW_MAGIC = 0x3243524e;     // 'NRC2'
+static const uint32_t NORCOW_MAGIC_V0 = 0x5743524e;  // NRCW
+static const uint32_t META_MAGIC_V10 = 0x525a5254;   // 'TRZR'
```
[the remark this answers is not in the document]
Comment (andrewkozlik) [reply legacy/intermediate_fw/trezor.c:40]: Yes this is correct. `NORCOW_MAGIC_V0` "NRCW" was used only on Trezor T before firmware 2.1.0.

**Review by andrewkozlik — CHANGES_REQUESTED** [review]
Please use [fixup commits](https://docs.trezor.io/trezor-firmware/misc/review.html#review-process). They make it easier to review.

File: legacy/intermediate_fw/trezor.c [hunk 2]
```diff
@@ -67,16 +74,15 @@ erase_firmware_and_storage(void) {
   FLASH_KEYR = FLASH_KEYR_KEY1;
   FLASH_KEYR = FLASH_KEYR_KEY2;
 
-  // Erase storage sectors to prevent firmware downgrade to vulnerable version
+  // Erase storage sectors
```
[the remark this answers is not in the document]
Comment (andrewkozlik) [reply legacy/intermediate_fw/trezor.c:77]: I think you are correct that we don't need to erase the storage here.

File: legacy/intermediate_fw/trezor.c [hunk 3]
```diff
@@ -112,14 +118,20 @@ int main(void) {
...
-               NULL, "DO NOT UNPLUG", "YOUR TREZOR!", NULL);
-  oledRefresh();
-
-  // from this point the execution is from RAM instead of flash
-  erase_fw_and_reboot();
+  check_and_replace_bootloader(false);
+
+  // storage is initialized
+  if (memcmp(FLASH_PTR(FLASH_META_START), &NORCOW_MAGIC,
+             sizeof(NORCOW_MAGIC)) == 0 ||
```
[the remark this answers is not in the document]
Comment (prusnak) [reply legacy/intermediate_fw/trezor.c]: Addressed in 8f9acc873e0da66d236be5f58fd83181a09f90b1

**Review by andrewkozlik — APPROVED** [review]
utACK

File: legacy/intermediate_fw/trezor.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (prusnak) [reply legacy/intermediate_fw/trezor.c]: Ok, addressed in another fixup - b4c7d22806a74cf3c073b85fb06f787d18baac83

**Review by andrewkozlik — APPROVED** [review]

**Review by onvej-sl — APPROVED** [review]

**Pull request merged**